### PR TITLE
Continued Record Selector UI implementation

### PR DESCRIPTION
**Pull request by seancolsen**
## Demo and notes

https://user-images.githubusercontent.com/42411/185246882-574e85fe-1080-4c84-9cce-552d21e26cc6.mp4

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by kgodey** [comment]
Nice work, @seancolsen! This looks great.

**Review by pavish — CHANGES_REQUESTED** [review]
@seancolsen I tested the functionality and it looks pretty good. I skimmed through the record selector specific code and did not review it in depth, I did not see any major concerns.

The review comments above are on common pieces of code and I'd like them resolved before merging in the PR, mainly addressing the removal of value binding.

File: mathesar_ui/src/component-library/debounce/Debounce.svelte
Comment (seancolsen) [reply mathesar_ui/src/component-library/debounce/Debounce.svelte:11]: Oversight on my part. Thanks for catching! Fixed in 13bd7334898c066c6d9d91dd0d866005fe3ad533.

File: mathesar_ui/src/components/cell-fabric/DynamicInput.svelte
Comment (seancolsen) [reply mathesar_ui/src/components/cell-fabric/DynamicInput.svelte:36]: Thanks for catching! I originally went a different direction with this PR and subsequently squashed my changes. Apparently this one slipped through. Whoops! Fixed in 8bef76da9e6c390bd60d8958e9de7614d51649f1.

File: mathesar_ui/src/component-library/debounce/Debounce.svelte
Comment (seancolsen) [reply mathesar_ui/src/component-library/debounce/Debounce.svelte]: Thanks for this suggestion. I didn't like this design in my PR either. I've added 26091ce179f2affc9256a0877fc1b21a65528d7d to improve it along the lines that you've suggested. The `Debounce` component is now decoupled from the concerns of `input` vs `artificialInput`. I wasn't able to make the `update` function quite as simple as your example, because the `Debounce` component needs to handle events differently depending on whether they should be debounced. I hope the design I went with will suffice

File: mathesar_ui/src/component-library/common/types/ArtificialEvents.ts
Comment (seancolsen) [reply mathesar_ui/src/component-library/common/types/ArtificialEvents.ts:4]: Yeah, I also really don't like the mess of `input` and `artificialInput`. I wrestled with this challenge quite a bit in this PR, trying a number of different approaches to debounce a dynamic input. I was hoping to find a design that wouldn't require modifying _all_ of our input components. I think this topic deserves a discussion via a sync call so that we can establish a good pattern to use throughout the component library. My opinion on the best design has been continuing to evolve throughout my work on this PR, but where I currently stand is that I'd like to use _props_ (instead of events) to handle imperative upwards data flow from all our input components. We've debated this before, but I think it's worth revisiting in the context of this specific use-case.

File: mathesar_ui/src/component-library/debounce/Debounce.svelte
Comment (pavish) [reply mathesar_ui/src/component-library/debounce/Debounce.svelte]: Looks good to me

File: mathesar_ui/src/component-library/common/types/ArtificialEvents.ts
Comment (pavish) [reply mathesar_ui/src/component-library/common/types/ArtificialEvents.ts:4]: Okay. I think we can have a frontend sync call for this during the cooldown. I'll leave it to you to schedule one.

**Review by pavish — APPROVED** [review]

**Pull request merged**
